packages/medusa/src/services/product-variant-inventory.ts

```diff
@@ -0,0 +1,376 @@
+    variantId: string
+  ): Promise<InventoryItemDTO[]> {
+    if (!this.inventoryService) {
+      return []
+    }
+
+    const variantInventory = await this.listByVariant(variantId)
+    const [items] = await this.inventoryService.listInventoryItems({
+      id: variantInventory.map((i) => i.inventory_item_id),
+    })
+
+    return items
+  }
+
+  async attachInventoryItem(
+    variantId: string,
+    inventoryItemId: string,
+    quantity?: number
+  ): Promise<ProductVariantInventoryItem> {
+    const manager = this.transactionManager_ || this.manager_
+
+    // Verify that variant exists
+    await this.productVariantService
+      .withTransaction(manager)
+      .retrieve(variantId, {
+        select: ["id"],
+      })
+
+    // Verify that item exists
+    await this.inventoryService.retrieveInventoryItem(inventoryItemId, {
+      select: ["id"],
+    })
+
+    const variantInventoryRepo = manager.getRepository(
+      ProductVariantInventoryItem
+    )
+
+    const existing = await variantInventoryRepo.findOne({
+      where: {
+        variant_id: variantId,
+        inventory_item_id: inventoryItemId,
+      },
+    })
+
+    if (existing) {
+      return existing
+    }
+
+    let quantityToStore = 1
+    if (typeof quantity !== "undefined") {
+      if (quantity < 1) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Quantity must be greater than 0"
+        )
+      } else {
+        quantityToStore = quantity
+      }
+    }
+
+    const variantInventory = variantInventoryRepo.create({
+      variant_id: variantId,
+      inventory_item_id: inventoryItemId,
+      quantity: quantityToStore,
+    })
+
+    return await variantInventoryRepo.save(variantInventory)
+  }
+
+  async detachInventoryItem(
+    variantId: string,
+    inventoryItemId: string
+  ): Promise<void> {
+    const manager = this.transactionManager_ || this.manager_
+
+    const variantInventoryRepo = manager.getRepository(
+      ProductVariantInventoryItem
+    )
+
+    const existing = await variantInventoryRepo.findOne({
+      where: {
+        variant_id: variantId,
+        inventory_item_id: inventoryItemId,
+      },
+    })
+
+    if (existing) {
+      await variantInventoryRepo.remove(existing)
+    }
+  }
+
+  async reserveQuantity(
+    variantId: string,
+    quantity: number,
+    options: {
+      location_id?: string
+      line_item_id?: string
+      sales_channel_id?: string | null
+    } = {}
+  ): Promise<void> {
+    const manager = this.transactionManager_ || this.manager_
+
+    if (!this.inventoryService) {
+      const variantServiceTx =
+        this.productVariantService.withTransaction(manager)
+      const variant = await variantServiceTx.retrieve(variantId, {
+        select: ["id", "inventory_quantity"],
+      })
+      await variantServiceTx.update(variant.id, {
+        inventory_quantity: variant.inventory_quantity - quantity,
+      })
+      return
```

Comment:
todo:
```suggestion
      return await variantServiceTx.update(variant.id, {
        inventory_quantity: variant.inventory_quantity - quantity,
      })
```

Reply:
This is not allowed due to the return type of the method being void and we need to terminate early in this case. Alternative is to wrap everything following this in an else statement. I think this is better though